resolver/changer_test.go

```diff
@@ -0,0 +1,99 @@
+package resolver
+
+import (
+	"testing"
+
+	mesos "github.com/mesos/mesos-go/mesosproto"
+)
+
+func TestMasterChanger(t *testing.T) {
+	newInfo := func(name string) *mesos.MasterInfo {
+		return &mesos.MasterInfo{Hostname: &name}
+	}
+
+	newMasters := func(masters []string) []*mesos.MasterInfo {
+		ms := make([]*mesos.MasterInfo, len(masters))
+
+		for i, m := range masters {
+			ms[i] = newInfo(m)
+		}
+
+		return ms
+	}
+
+	mc := NewMasterChanger([]string{"b:5050", "a:5050"})
+
+	onMasterChanged := func(name string) func() {
+		return func() {
+			mc.OnMasterChanged(newInfo(name))
+		}
+	}
+
+	updatedMasters := func(masters []string) func() {
+		return func() {
+			mc.UpdatedMasters(newMasters(masters))
+		}
+	}
+
+	var tests = []struct {
+		f       func()
+		masters []string
+	}{
+		{
+			func() { mc.Emit() },
+			[]string{"", "b:5050", "a:5050"},
+		},
+		{
+			onMasterChanged("a"),
+			[]string{"a:5050", "b:5050"},
+		},
+		{
+			onMasterChanged("b"),
+			[]string{"b:5050"},
+		},
+		{
+			updatedMasters([]string{"c", "d", "b"}),
+			[]string{"b:5050", "c:5050", "d:5050"},
+		},
+		{
+			onMasterChanged(""),
+			[]string{"0.0.0.0:5050", "c:5050", "d:5050"},
+		},
+		{
+			updatedMasters([]string{}),
+			[]string{"0.0.0.0:5050"},
+		},
+		{
+			func() { mc.UpdatedMasters(nil) },
+			nil,
+		},
+		{
+			func() { mc.OnMasterChanged(nil) },
+			nil,
+		},
+	}
+
+	stop := make(chan struct{})
+	go func() {
+		for _, tt := range tests {
+			tt.f()
+		}
+		stop <- struct{}{}
+	}()
+
+	i := 0
+	for {
+		select {
+		case ms := <-mc.ChangedChan():
```

Comment:
If we parameterize the channel capacity on `MasterChanged` construction, we wouldn't need all this jazz of go-routines.